x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeListener.java

```diff
@@ -73,13 +117,75 @@ ActionListener<Void> acquireAvoid() {
     /**
      * Acquires a new listener that collects compute result. This listener will also collects warnings emitted during compute
      */
-    ActionListener<ComputeResponse> acquireCompute() {
+    ActionListener<ComputeResponse> acquireComputeForDataNodes() {
+        return acquireAvoid().map(resp -> {
+            responseHeaders.collect();
+            var profiles = resp.getProfiles();
+            if (profiles != null && profiles.isEmpty() == false) {
+                collectedProfiles.addAll(profiles);
+            }
+            return null;
+        });
+    }
+
+    /**
+     * The ActionListener to be used on the coordinating cluster when sending a cross-cluster
+     * compute request to a remote cluster.
+     * @param clusterAlias clusterAlias of cluster receiving the remote compute request
+     * @return Listener that will fill in all metadata from to remote cluster into the
+     *         {@link EsqlExecutionInfo}  for the clusterAlias cluster.
+     */
+    ActionListener<ComputeResponse> acquireCCSCompute(String clusterAlias) {
+        assert clusterAlias != null : "Must provide non-null cluster alias to acquireCompute";
+        assert esqlExecutionInfo != null : "When providing cluster alias to acquireCompute, EsqlExecutionInfo must not be null";
+        return acquireAvoid().map(resp -> {
+            responseHeaders.collect();
+            var profiles = resp.getProfiles();
+            if (profiles != null && profiles.isEmpty() == false) {
+                collectedProfiles.addAll(profiles);
+            }
+            esqlExecutionInfo.swapCluster(
+                clusterAlias,
+                (k, v) -> new EsqlExecutionInfo.Cluster.Builder(v)
+                    // MP TODO: if we get here does that mean that the remote search is finished and was SUCCESSFUL?
+                    // for now ESQL doesn't return partial results, so set status to SUCCESSFUL
+                    .setStatus(EsqlExecutionInfo.Cluster.Status.SUCCESSFUL)
+                    .setTook(resp.getTook())
+                    .setTotalShards(resp.getTotalShards())
+                    .setSuccessfulShards(resp.getSuccessfulShards())
+                    .setSkippedShards(resp.getSkippedShards())
+                    .setFailedShards(resp.getFailedShards())
+                    .build()
+            );
+            return null;
+        });
+    }
+
+    /**
+     * Acts like {@code acquireCompute} for handling the response(s) form the runComputeOnDataNodes
+     * phase. Per-cluster took time is recorded in the {@link EsqlExecutionInfo}.
+     * @param clusterAlias remote cluster alias the data node compute is running on
+     */
+    ActionListener<ComputeResponse> acquireComputeForDataNodes(String clusterAlias, Configuration configuration) {
```

Comment:
What is the difference between acquireComputeForDataNodes with and without arguments? Can we streamline them?

Reply:
Two part answer:

First, I didn't mean to rename the original `acquireCompute`. I see now that IntelliJ grabbed both when I did a rename of the one that takes args. I renamed it back in [this commit](https://github.com/elastic/elasticsearch/pull/112472/commits/1aea20f6e7043ed3f99d112517ba0989efe6f28f).

Second, you might still ask - why do we need `acquireCompute`, `acquireComputeForDataNodes` and `acquireCCSCompute` now. The reason is that `acquireCompute` was used at various levels of the "compute" phases and sub-phases (I'm not sure the right ESQL terminology for this yet). I need some spot or hook where I know that higher level results are coming back - either the final response from a remote cluster (`acquireCCSCompute`) or each "final" result from each data node - so this is set at the top of the "compute on data nodes phase" (starts in `startComputeOnDataNodes`) in order to be able to calculate per-cluster took time. 

All the other "sub-phases" or sub-computes use the original `acquireCompute`, and those phases (AFAICT) hold no useful information for me in terms of gathering execution metadata.